modules/programs/firefox.nix

```diff
@@ -444,6 +532,124 @@ in {
             mkUserJs profile.settings profile.extraConfig profile.bookmarks;
         };
 
+      "${profilesPath}/${profile.path}/search.json.mozlz4" = mkIf
+        (profile.search.default != null || length profile.search.order != [ ]
+          || profile.search.engines != { }) {
+            force = profile.search.force;
+            source = let
+              settings = {
+                version = 6;
+
+                engines = let
+                  allEngines = (profile.search.engines //
+                    # If search.default isn't in search.engines, assume it's
+                    # app provided and include it in the set of all engines
+                    optionalAttrs (profile.search.default != null
+                      && !(hasAttr profile.search.default
+                        profile.search.engines)) {
+                          ${profile.search.default} = { };
+                        });
+
+                  # Map allEngines to a list and order by search.order
+                  orderedEngineList = (imap (order: name:
+                    let engine = (allEngines.${name} or { });
+                    in engine // {
+                      inherit name;
+                      metaData = (engine.metaData or { }) // { inherit order; };
+                    }) profile.search.order) ++ (mapAttrsToList
+                      (name: config: config // { inherit name; })
+                      (removeAttrs allEngines profile.search.order));
+
+                  engines = map (config:
+                    let
+                      name = config.name;
+                      isAppProvided = (removeAttrs config [ "name" "metaData" ])
+                        == { };
+                      metaData = config.metaData or { };
+                    in mapAttrs' (name: value: {
+                      # Map nice field names to internal field names.
+                      # This is intended to be exhaustive, but any future fields
+                      # will either have to be specified with an underscore,
+                      # or added to this map.
+                      name = {
+                        name = "_name";
+                        isAppProvided = "_isAppProvided";
+                        loadPath = "_loadPath";
+                        hasPreferredIcon = "_hasPreferredIcon";
+                        searchForm = "__searchForm";
+                        updateInterval = "_updateInterval";
+                        updateURL = "_updateURL";
+                        iconUpdateURL = "_iconUpdateURL";
+                        iconURL = "_iconURL";
+                        iconMapObj = "_iconMapObj";
+                        metaData = "_metaData";
+                        orderHint = "_orderHint";
+                        definedAliases = "_definedAliases";
+                        urls = "_urls";
+                      }.${name} or name;
```

Comment:
Perhaps this could be done prettier using `genAttrs`.

Reply:
Oh yep! That could make things simpler, but searchForm requires two underscores so it would still be an exception.